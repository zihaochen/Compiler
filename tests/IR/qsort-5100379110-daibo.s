.data
	_printf_cnt: .space 4
	_a: .space 4
	_n: .space 4
	ASU_1: .space 40404
	str_1: .space 16
	str_2: .space 8
.text
main:
	move $fp, $sp     # start using memory here
	move $gp, $sp     # set global pointer (unused)
	subu $sp, $sp, 4
	la $t0, ASU_1
	sw $t0, _a
	li $t0, 0
	sw $t0, _printf_cnt
	li $t0, 1000
	sw $t0, _n
	li $t0, 37
	la $t1, str_1
	sw $t0, 0($t1)
	li $t0, 100
	la $t1, str_1
	sw $t0, 4($t1)
	li $t0, 32
	la $t1, str_1
	sw $t0, 8($t1)
	li $t0, 10
	la $t1, str_2
	sw $t0, 0($t1)
	j _main
_qsrt:
	sw $ra, ($sp)
	lw $t0, 8($sp)
	sw $t0, -4($sp)
	lw $t0, 4($sp)
	sw $t0, -8($sp)
	lw $t2, 4($sp)
	lw $t1, 8($sp)
	add $t0, $t1, $t2
	sw $t0, -16($sp)
	li $t2, 2
	lw $t1, -16($sp)
	div $t0, $t1, $t2
	sw $t0, -20($sp)
	li $t0, 0
	sw $t0, -24($sp)
	li $t2, 4
	lw $t1, -20($sp)
	mul $t0, $t1, $t2
	sw $t0, -28($sp)
	lw $t2, -28($sp)
	lw $t1, -24($sp)
	add $t0, $t1, $t2
	sw $t0, -24($sp)
	lw $t2, -24($sp)
	lw $t1, _a
	add $t0, $t1, $t2
	sw $t0, -32($sp)
	lw $t0, -32($sp)
	lw $t1, ($t0)
	sw $t1, -36($sp)
	lw $t0, -36($sp)
	sw $t0, -12($sp)
L2:
	lw $t2, -8($sp)
	lw $t1, -4($sp)
	sle $t0, $t1, $t2
	sw $t0, -40($sp)
	lw $t0, -40($sp)
	bne $t0, $0, L3
	b L1
L3:
L5:
	li $t0, 0
	sw $t0, -44($sp)
	li $t2, 4
	lw $t1, -4($sp)
	mul $t0, $t1, $t2
	sw $t0, -48($sp)
	lw $t2, -48($sp)
	lw $t1, -44($sp)
	add $t0, $t1, $t2
	sw $t0, -44($sp)
	lw $t2, -44($sp)
	lw $t1, _a
	add $t0, $t1, $t2
	sw $t0, -52($sp)
	lw $t0, -52($sp)
	lw $t1, ($t0)
	sw $t1, -56($sp)
	lw $t2, -12($sp)
	lw $t1, -56($sp)
	slt $t0, $t1, $t2
	sw $t0, -60($sp)
	lw $t0, -60($sp)
	bne $t0, $0, L6
	b L4
L6:
	li $t2, 1
	lw $t1, -4($sp)
	add $t0, $t1, $t2
	sw $t0, -72($sp)
	lw $t0, -72($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -68($sp)
	li $t2, 1
	lw $t1, -68($sp)
	sub $t0, $t1, $t2
	sw $t0, -64($sp)
	b L5
L4:
L8:
	li $t0, 0
	sw $t0, -76($sp)
	li $t2, 4
	lw $t1, -8($sp)
	mul $t0, $t1, $t2
	sw $t0, -80($sp)
	lw $t2, -80($sp)
	lw $t1, -76($sp)
	add $t0, $t1, $t2
	sw $t0, -76($sp)
	lw $t2, -76($sp)
	lw $t1, _a
	add $t0, $t1, $t2
	sw $t0, -84($sp)
	lw $t0, -84($sp)
	lw $t1, ($t0)
	sw $t1, -88($sp)
	lw $t2, -12($sp)
	lw $t1, -88($sp)
	sgt $t0, $t1, $t2
	sw $t0, -92($sp)
	lw $t0, -92($sp)
	bne $t0, $0, L9
	b L7
L9:
	li $t2, 1
	lw $t1, -8($sp)
	sub $t0, $t1, $t2
	sw $t0, -104($sp)
	lw $t0, -104($sp)
	sw $t0, -8($sp)
	lw $t0, -8($sp)
	sw $t0, -100($sp)
	li $t2, 1
	lw $t1, -100($sp)
	add $t0, $t1, $t2
	sw $t0, -96($sp)
	b L8
L7:
	lw $t2, -8($sp)
	lw $t1, -4($sp)
	sle $t0, $t1, $t2
	sw $t0, -108($sp)
	lw $t0, -108($sp)
	bne $t0, $0, L11
	b L10
L11:
	li $t0, 0
	sw $t0, -116($sp)
	li $t2, 4
	lw $t1, -4($sp)
	mul $t0, $t1, $t2
	sw $t0, -120($sp)
	lw $t2, -120($sp)
	lw $t1, -116($sp)
	add $t0, $t1, $t2
	sw $t0, -116($sp)
	lw $t2, -116($sp)
	lw $t1, _a
	add $t0, $t1, $t2
	sw $t0, -124($sp)
	lw $t0, -124($sp)
	lw $t1, ($t0)
	sw $t1, -128($sp)
	lw $t0, -128($sp)
	sw $t0, -112($sp)
	li $t0, 0
	sw $t0, -136($sp)
	li $t2, 4
	lw $t1, -8($sp)
	mul $t0, $t1, $t2
	sw $t0, -140($sp)
	lw $t2, -140($sp)
	lw $t1, -136($sp)
	add $t0, $t1, $t2
	sw $t0, -136($sp)
	lw $t2, -136($sp)
	lw $t1, _a
	add $t0, $t1, $t2
	sw $t0, -144($sp)
	lw $t0, -144($sp)
	lw $t1, ($t0)
	sw $t1, -148($sp)
	li $t0, 0
	sw $t0, -152($sp)
	li $t2, 4
	lw $t1, -4($sp)
	mul $t0, $t1, $t2
	sw $t0, -156($sp)
	lw $t2, -156($sp)
	lw $t1, -152($sp)
	add $t0, $t1, $t2
	sw $t0, -152($sp)
	lw $t2, -152($sp)
	lw $t1, _a
	add $t0, $t1, $t2
	sw $t0, -160($sp)
	lw $t0, -160($sp)
	lw $t1, -148($sp)
	sw $t1, ($t0)
	lw $t0, -160($sp)
	lw $t1, ($t0)
	sw $t1, -132($sp)
	lw $t0, -160($sp)
	lw $t1, ($t0)
	sw $t1, -132($sp)
	li $t0, 0
	sw $t0, -172($sp)
	li $t2, 4
	lw $t1, -8($sp)
	mul $t0, $t1, $t2
	sw $t0, -176($sp)
	lw $t2, -176($sp)
	lw $t1, -172($sp)
	add $t0, $t1, $t2
	sw $t0, -172($sp)
	lw $t2, -172($sp)
	lw $t1, _a
	add $t0, $t1, $t2
	sw $t0, -180($sp)
	lw $t0, -180($sp)
	lw $t1, -112($sp)
	sw $t1, ($t0)
	lw $t0, -180($sp)
	lw $t1, ($t0)
	sw $t1, -168($sp)
	lw $t0, -180($sp)
	lw $t1, ($t0)
	sw $t1, -168($sp)
	li $t2, 1
	lw $t1, -4($sp)
	add $t0, $t1, $t2
	sw $t0, -196($sp)
	lw $t0, -196($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -192($sp)
	li $t2, 1
	lw $t1, -192($sp)
	sub $t0, $t1, $t2
	sw $t0, -188($sp)
	li $t2, 1
	lw $t1, -8($sp)
	sub $t0, $t1, $t2
	sw $t0, -208($sp)
	lw $t0, -208($sp)
	sw $t0, -8($sp)
	lw $t0, -8($sp)
	sw $t0, -204($sp)
	li $t2, 1
	lw $t1, -204($sp)
	add $t0, $t1, $t2
	sw $t0, -200($sp)
L10:
	b L2
L1:
	lw $t2, -8($sp)
	lw $t1, 8($sp)
	slt $t0, $t1, $t2
	sw $t0, -212($sp)
	lw $t0, -212($sp)
	bne $t0, $0, L13
	b L12
L13:
	lw $t0, 8($sp)
	sw $t0, -228($sp)
	lw $t0, -8($sp)
	sw $t0, -232($sp)
	subu $sp, $sp, 236
	jal _qsrt
	addi $sp, $sp, 236
	sw $v0, -216($sp)
	lw $ra, ($sp)
L12:
	lw $t2, 4($sp)
	lw $t1, -4($sp)
	slt $t0, $t1, $t2
	sw $t0, -220($sp)
	lw $t0, -220($sp)
	bne $t0, $0, L15
	b L14
L15:
	lw $t0, -4($sp)
	sw $t0, -228($sp)
	lw $t0, 4($sp)
	sw $t0, -232($sp)
	subu $sp, $sp, 236
	jal _qsrt
	addi $sp, $sp, 236
	sw $v0, -224($sp)
	lw $ra, ($sp)
L14:
	li $v0, 0
	sw $v0, 0($sp)
	jr $ra
	jr $ra
_main:
	sw $ra, ($sp)
	la $t0, str_1
	sw $t0, -68($sp)
	la $t0, str_2
	sw $t0, -104($sp)
	li $t0, 1
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -8($sp)
L17:
	lw $t2, _n
	lw $t1, -4($sp)
	sle $t0, $t1, $t2
	sw $t0, -12($sp)
	lw $t0, -12($sp)
	bne $t0, $0, L18
	b L16
L18:
	li $t2, 1
	lw $t1, _n
	add $t0, $t1, $t2
	sw $t0, -20($sp)
	lw $t2, -4($sp)
	lw $t1, -20($sp)
	sub $t0, $t1, $t2
	sw $t0, -24($sp)
	li $t0, 0
	sw $t0, -28($sp)
	li $t2, 4
	lw $t1, -4($sp)
	mul $t0, $t1, $t2
	sw $t0, -32($sp)
	lw $t2, -32($sp)
	lw $t1, -28($sp)
	add $t0, $t1, $t2
	sw $t0, -28($sp)
	lw $t2, -28($sp)
	lw $t1, _a
	add $t0, $t1, $t2
	sw $t0, -36($sp)
	lw $t0, -36($sp)
	lw $t1, -24($sp)
	sw $t1, ($t0)
	lw $t0, -36($sp)
	lw $t1, ($t0)
	sw $t1, -16($sp)
	lw $t0, -36($sp)
	lw $t1, ($t0)
	sw $t1, -16($sp)
	li $t2, 1
	lw $t1, -4($sp)
	add $t0, $t1, $t2
	sw $t0, -52($sp)
	lw $t0, -52($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -48($sp)
	li $t2, 1
	lw $t1, -48($sp)
	sub $t0, $t1, $t2
	sw $t0, -44($sp)
	b L17
L16:
	li $t0, 1
	sw $t0, -112($sp)
	lw $t0, _n
	sw $t0, -116($sp)
	subu $sp, $sp, 120
	jal _qsrt
	addi $sp, $sp, 120
	sw $v0, -56($sp)
	lw $ra, ($sp)
	li $t0, 1
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -60($sp)
L20:
	lw $t2, _n
	lw $t1, -4($sp)
	sle $t0, $t1, $t2
	sw $t0, -64($sp)
	lw $t0, -64($sp)
	bne $t0, $0, L21
	b L19
L21:
	li $t0, 0
	sw $t0, -72($sp)
	li $t2, 4
	lw $t1, -4($sp)
	mul $t0, $t1, $t2
	sw $t0, -76($sp)
	lw $t2, -76($sp)
	lw $t1, -72($sp)
	add $t0, $t1, $t2
	sw $t0, -72($sp)
	lw $t2, -72($sp)
	lw $t1, _a
	add $t0, $t1, $t2
	sw $t0, -80($sp)
	lw $t0, -80($sp)
	lw $t1, ($t0)
	sw $t1, -84($sp)
	lw $t0, -68($sp)
	sw $t0, -112($sp)
	lw $t0, -84($sp)
	sw $t0, -116($sp)
	li $t0, 2
	sw $t0, _printf_cnt
	subu $sp, $sp, 120
	jal _printf
	addi $sp, $sp, 120
	sw $v0, -88($sp)
	lw $ra, ($sp)
	li $t2, 1
	lw $t1, -4($sp)
	add $t0, $t1, $t2
	sw $t0, -100($sp)
	lw $t0, -100($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -96($sp)
	li $t2, 1
	lw $t1, -96($sp)
	sub $t0, $t1, $t2
	sw $t0, -92($sp)
	b L20
L19:
	lw $t0, -104($sp)
	sw $t0, -112($sp)
	li $t0, 1
	sw $t0, _printf_cnt
	subu $sp, $sp, 116
	jal _printf
	addi $sp, $sp, 116
	sw $v0, -108($sp)
	lw $ra, ($sp)
	li $v0, 0
	sw $v0, 0($sp)
	jr $ra
	jr $ra

_printf:
	lw $t0, _printf_cnt
	mul $t0, $t0, 4
	add $t0, $t0, $sp
	lw $a1, ($t0)
	move $a2, $t0
		
_printf_loop:
	lb $a0, 0($a1)
	beq $a0, 0, _printf_end
	addu $a1, $a1, 4
	beq $a0, '%', _printf_fmt
	li $v0, 11
	syscall
	b _printf_loop

_printf_fmt:
	lb $a0, 0($a1)
	addu $a1, $a1, 4
	beq $a0, 'd', _printf_int
	beq $a0, 's', _printf_str
	beq $a0, 'c', _printf_char
	beq $a0, '0', _printf_width
	beq $a0, '.', _printf_width

_printf_int:
	subu $a2, $a2, 4
	lw $a0, 0($a2)
	li $v0, 1
	syscall
	b _printf_loop

_printf_str:
	subu $a2, $a2, 4
	lw $a0, 0($a2)
	li $v0, 4
	syscall
	b _printf_loop

_printf_char:
	subu $a2, $a2, 4
	lw $a0, 0($a2)
	li $v0, 11
	syscall
	b _printf_loop

_printf_width:
	lb $t1, 0($a1)
	addu $a1, $a1, 4
	sub $t1, $t1, '1'
	addu $a1, $a1, 4
	subu $a2, $a2, 4
	lw $t0, 0($a2)
	blt $t0, 10, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 100, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 1000, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 10000, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 100000, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 1000000, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 10000000, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 100000000, _printf_width_tmp
	subu $t1, $t1, 1

_printf_width_tmp:
	li $a0, '0'
	beq $t1, 0, _printf_width_tmp2
	subu $t1, $t1, 1
	li $v0, 11
	syscall
	b _printf_width_tmp

_printf_width_tmp2:
	move $a0, $t0
	li $v0, 1
	syscall
	b _printf_loop

_printf_end:
	jr $ra

_getchar:
	li $v0, 12
	syscall
	jr $ra

_malloc_struct:
	lw $a0, 4($sp)
	li $v0, 9
	syscall
	move $t0, $v0
	li $v0, 9
	li $a0, 4
	syscall
	sw $t0, ($v0)
	jr $ra

_malloc:
	lw $a0, 4($sp)
	li $v0, 9
	syscall
	jr $ra

