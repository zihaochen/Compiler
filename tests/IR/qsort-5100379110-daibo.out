n = 10000
i = l
j = r
Temp2294 = l ADD r
Temp2295 = Temp2294 DIV 2
Temp2296 = 0
Temp2297 = Temp2295 MUL 4
Temp2296 = Temp2296 ADD Temp2297
Temp2298 = a ADD Temp2296
Temp2299 = *Temp2298, 4
x = Temp2299
Label :415	Temp2300 = i LE j
If Not Equal to Zero Temp2300 Goto Label :416	
Goto label:414
Label :416	Label :418	Temp2301 = 0
Temp2302 = i MUL 4
Temp2301 = Temp2301 ADD Temp2302
Temp2303 = a ADD Temp2301
Temp2304 = *Temp2303, 4
Temp2305 = Temp2304 LT x
If Not Equal to Zero Temp2305 Goto Label :419	
Goto label:417
Label :419	Temp2306 = i
i = i ADD 1
Goto label:418
Label :417	Label :421	Temp2307 = 0
Temp2308 = j MUL 4
Temp2307 = Temp2307 ADD Temp2308
Temp2309 = a ADD Temp2307
Temp2310 = *Temp2309, 4
Temp2311 = Temp2310 GT x
If Not Equal to Zero Temp2311 Goto Label :422	
Goto label:420
Label :422	Temp2312 = j
j = j SUB 1
Goto label:421
Label :420	Temp2313 = i LE j
If Not Equal to Zero Temp2313 Goto Label :424	
Goto label:425
Label :424	Temp2314 = 0
Temp2315 = i MUL 4
Temp2314 = Temp2314 ADD Temp2315
Temp2316 = a ADD Temp2314
Temp2317 = *Temp2316, 4
temp = Temp2317
Temp2319 = 0
Temp2320 = j MUL 4
Temp2319 = Temp2319 ADD Temp2320
Temp2321 = a ADD Temp2319
Temp2322 = *Temp2321, 4
Temp2323 = 0
Temp2324 = i MUL 4
Temp2323 = Temp2323 ADD Temp2324
Temp2325 = a ADD Temp2323
*Temp2325 = Temp2322 , 4
Temp2328 = 0
Temp2329 = j MUL 4
Temp2328 = Temp2328 ADD Temp2329
Temp2330 = a ADD Temp2328
*Temp2330 = temp , 4
Temp2332 = i
i = i ADD 1
Temp2333 = j
j = j SUB 1
Goto label:423
Label :425	Label :423	Goto label:415
Label :414	Temp2334 = l LT j
If Not Equal to Zero Temp2334 Goto Label :427	
Goto label:428
Label :427	BasicParam l
BasicParam j
call qsrt , 2
Goto label:426
Label :428	Label :426	Temp2336 = i LT r
If Not Equal to Zero Temp2336 Goto Label :430	
Goto label:431
Label :430	BasicParam i
BasicParam r
call qsrt , 2
Goto label:429
Label :431	Label :429	Return BasicParam 0

i = 1
Temp2338 = i
Label :433	Temp2339 = i LE n
If Not Equal to Zero Temp2339 Goto Label :434	
Goto label:432
Label :434	Temp2341 = n ADD 1
Temp2342 = Temp2341 SUB i
Temp2343 = 0
Temp2344 = i MUL 4
Temp2343 = Temp2343 ADD Temp2344
Temp2345 = a ADD Temp2343
*Temp2345 = Temp2342 , 4
Temp2347 = i
i = i ADD 1
Goto label:433
Label :432	BasicParam 1
BasicParam n
call qsrt , 2
i = 1
Temp2349 = i
Label :436	Temp2350 = i LE n
If Not Equal to Zero Temp2350 Goto Label :437	
Goto label:435
Label :437	Temp2351 = &"%d "
Temp2352 = 0
Temp2353 = i MUL 4
Temp2352 = Temp2352 ADD Temp2353
Temp2354 = a ADD Temp2352
Temp2355 = *Temp2354, 4
BasicParam Temp2351
BasicParam Temp2355
call printf , 2
Temp2357 = i
i = i ADD 1
Goto label:436
Label :435	Temp2358 = &"
"
BasicParam Temp2358
call printf , 1
Return BasicParam 0

